# Preface

Analysis_of_peak_downstream are consist of three parts

-   Genome Coordinate Conversion,

-   Peak overlap

-   peak Annotation, Comparison and Visualization

<br>

------------------------------------------------------------------------

## Genome Coordinate Conversion Preface

When conducting research, we need to unify the genome coordinate system. The process of converting from one version of the coordinate system to another is called **liftover(UCSC)** or **crossmap (Ensembl)**.

-   The **UCSC and Ensembl** provides a ready-to-use **web version**:
    * [**UCSC** web version](https://genome.ucsc.edu/cgi-bin/hgLiftOver)
    * [**Ensembl** web version](https://grch37.ensembl.org/Homo_sapiens/Tools/AssemblyConverter) **(Only support species: human, mouse, Caenorhabditis elegans, Saccharomyces cerevisiae, Zebrafish)**

<br>

------------------------------------------------------------------------

## Peak overlap Preface

This part about how to find overlaps of peaks?

-   First, we use the peak files to create `GRanges` objects.
<br>
-   `findOverlaps()` : Creating `NCList` and `GNCList` objects to find the overlaps between two "range-based" objects.
<br>
-   `findOverlapsOfPeaks()` :Ising `findOverlaps()` another version for peaks.
<br>
-   `makeVennDiagram()` : Making Venn Diagram for peaks.

<br>

------------------------------------------------------------------------

## Peak Annotation, Comparison and Visualization Preface

This part about [`ChIPseeker`](https://bioconductor.org/packages/release/bioc/vignettes/ChIPseeker/inst/doc/ChIPseeker.html) usage, an R package for **peak Annotation, Comparison and Visualization**

-   If two ChIP Seq data, obtained by two different binding proteins, overlap significantly, these two proteins may form a complex or have interaction in regulation chromosome remodelling or gene expression.Â 
<br>
-   [`ChIPseeker`](https://bioconductor.org/packages/release/bioc/vignettes/ChIPseeker/inst/doc/ChIPseeker.html) support statistical testing of significant overlap among ChIP seq data sets, and incorporate open access database GEO (contains **17,000 bed file**) for users to compare their own dataset to those deposited in database
<br>
-   Converting genome coordinations from one genome version to another is also supported

<br>

------------------------------------------------------------------------
# Basic knowledge

## 0-base or 1-baseï¼Ÿ

-   [è¯¦è§](https://genome-blog.gi.ucsc.edu/blog/2016/12/12/the-ucsc-genome-browser-coordinate-counting-systems/)<br>
-   åœ¨**0-based**åæ ‡ç³»ä¸­ï¼Œèµ·å§‹ä½ç½®ä» 0 å¼€å§‹è®¡æ•°ï¼ŒèŒƒå›´æ˜¯ **[start, end)** ï¼Œå³ **end ä½ç½®ä¸åŒ…å«åœ¨åŒºé—´å†…**ã€‚
-   åœ¨**1-based**åæ ‡ç³»ä¸­ï¼Œèµ·å§‹ä½ç½®ä» 1 å¼€å§‹è®¡æ•°ï¼ŒèŒƒå›´æ˜¯ **[start, end]** ï¼Œå³ **end ä½ç½®åŒ…å«åœ¨åŒºé—´å†…**ã€‚

| æ–‡ä»¶æ ¼å¼ | åæ ‡ç³»ç»Ÿ | è¯´æ˜ |
|------------------------|------------------------|------------------------|
| **BED (UCSC)** | **0-based** | `[start, end)` |
| **GTF/GFF (Ensembl)** | **1-based** | `[start, end]` |
| **VCF (Ensembl)** | **1-based** | å˜å¼‚ä½ç‚¹åæ ‡ä» 1 å¼€å§‹ï¼Œä¸æ˜¯åŒºé—´ï¼Œè€Œæ˜¯å•ä¸ªç¢±åŸºçš„ä½ç½® |
| **BAM/SAM** | **1-based** | æ¯”å¯¹èµ·å§‹ä½ç½®ä¸º 1-basedï¼ŒCIGAR æ“ä½œçš„ç»“æœç¡®å®šç»ˆæ­¢ä½ç½® |
| **LiftOver** | **0-based** | è¾“å…¥æ–‡ä»¶é€šå¸¸æ˜¯ 0-basedï¼ˆå¦‚ BEDï¼‰ |
| **CrossMap** | **æ ¹æ®è¾“å…¥æ ¼å¼** | æ”¯æŒå¤šç§æ ¼å¼ï¼Œå¦‚ BEDï¼ˆ0-basedï¼‰ã€GTF/GFF/VCFï¼ˆ1-basedï¼‰ |
<br>

------------------------------------------------------------------------
## GRCh38 vs hg38 or mRatBN7.2 vs rn7?

### GRCh38 vs hg38

-   **GRCh38:** The Human Reference Genome maintained by the Genome Reference Consortium (GRC) is the 38th edition of the Genome Reference Consortium Human series.
-   **hg38:** indicates the GRCh38 version used by UCSC

<br>

**Same**ğŸ˜€ğŸ˜€
* GRCh38 and hg38 share the **same genome sequence**, so **most of the coordinates are identical**

<br>

 **Different**ğŸ¥¸ğŸ¥¸
* **UCSC** uses the **UCSC-specific mitochondrial genome (chrM)** in **hg38**, while **GRCh38** may use a **different mitochondrial version** (such as **rCRS or GRCh38_MT**).
* **UCSC** has added **alternate haplotypes** in **hg38**

<br>

### What is the Haplotype Patches or alternate haplotypes ?

-   **å•å€å‹è¡¥ä¸ï¼ˆHaplotype Patchesï¼‰**

    ``` text
    chr1    1000000 1000500 region_A    100 +
    chr1    2000000 2000600 region_B    200 -
    chr1_KI270706v1_random  50000   50500   patch_region    150 +
    ```

    -   **chr1_KI270706v1_random** æ˜¯ å•å€å‹è¡¥ä¸ï¼Œç”¨äºä¿®å¤ chr1 ä¸ŠæŸä¸ªå¤æ‚åŒºåŸŸçš„é—®é¢˜ã€‚

    -   è¿™äº›è¡¥ä¸çš„åæ ‡é€šå¸¸æ˜¯ç‹¬ç«‹çš„ï¼Œå®ƒä»¬ä¸ä¼šç›´æ¥ä¸ä¸»æŸ“è‰²ä½“åæ ‡ç³»å¯¹åº”ï¼ˆå³ chr1 å’Œ chr1_KI270706v1_random å¯èƒ½æ˜¯ç›¸äº’ç‹¬ç«‹çš„ï¼‰

<br>

-   **æ›¿ä»£å•å€å‹ï¼ˆAlternate Haplotypesï¼‰**

    ``` text
    chr6    3000000 3000500 MHC_region_ref  500 +
    chr6_GL000256v2_alt 3000000 3000500 MHC_region_alt  500 +
    ```

    -   **chr6** æ˜¯æ ‡å‡†æŸ“è‰²ä½“çš„ MHC ï¼ˆMajor Histocompatibility Complex, ä¸»è¦ç»„ç»‡ç›¸å®¹æ€§å¤åˆä½“ï¼‰(è´Ÿè´£ç¼–ç ç»†èƒè¡¨é¢çš„åˆ†å­ï¼Œç”¨äºæŠ—åŸå‘ˆé€’å’Œå…ç–«ç³»ç»Ÿè¯†åˆ«)ï¼ˆä¸»å‚è€ƒåºåˆ—ï¼‰ã€‚

    -   **chr6_GL000256v2_alt** æ˜¯ MHC æ›¿ä»£å•å€å‹ï¼ˆalt haplotypeï¼‰ï¼Œç”¨äºæä¾›è¯¥åŒºåŸŸçš„å¦ä¸€ç§ç‰ˆæœ¬ï¼Œè¡¨ç¤ºåœ¨ä¸åŒäººç¾¤ä¸­å¯èƒ½å­˜åœ¨çš„å˜å¼‚ã€‚

<br>

### mRatBN7.2 vs rn7

It's the same version of the genome, just from a different database

-   **mRatBN7.2** from Ensemble

-   **rn7** from UCSC

<br>

### Conclusion

| **ç‰¹æ€§** | **mRatBN7.2** | **rn7** | **GRCh38** | **hg18** |
|---------------|---------------|---------------|---------------|---------------|
| **æ•°æ®æ¥æº** | Wellcome Sanger Institute / Ensembl | UCSC Genome Browser | Genome Reference Consortium (GRC) | UCSC Genome Browser |
| **å‘å¸ƒæœºæ„** | Ensembl | UCSC | GRC / Ensembl | UCSC |
| **æŸ“è‰²ä½“å‘½å** | 1, 2, X, Y, MT | chr1, chr2, chrX, chrY, chrM | 1, 2, X, Y, MT | chr1, chr2, chrX, chrY, chrM |
| **åæ ‡ç³»ç»Ÿ** | 1-based | 1-based | 1-based | 1-based |
| **å•å€å‹è¡¥ä¸ (Alternate Haplotypes)** | æœ‰ï¼ˆEnsembl ä¹Ÿä½¿ç”¨ï¼‰ | UCSC å¯èƒ½æœ‰ä¸åŒçš„å‘½åæ–¹å¼ | æœ‰ï¼ˆä¸ Ensembl ä¸€è‡´ï¼‰ | æ—  |

<br>

------------------------------------------------------------------------
## GRanges and IRanges objects

The central feature of **GenomicRanges** is the `GRanges` and `IRanges`(IntegerRanges)** object, which represents genomic intervals.

### GRanges objects

-   Each `GRanges` object can include the following

    -   components:Chromosome Names (`seqnames`)

    -   `Start` and `End` Positions (`ranges`)

    -   `Strand` Information

        -   Positive strand: `+`

        -   Negative strand: `-`

        -   Unspecified strand: `*`

    -   Additional Metadata(`score`, `Annotations`,`Other custom attributes`)

<BR>

**Create the `GRanges` object**

```R
# loading package
# processing genomic data.
library(rtracklayer)

# create two GRanges object
query <- GRanges(seqnames = c("chr1","chr1","chr1"),
                  ranges = IRanges(start = c(100, 300, 500), 
                                   end = c(200, 400, 600),
                                   names=c("a","b","c")),
                  score = c(10, 20, 30),
                  strand="+")

subject <- GRanges(seqnames = c("chr1", "chr1", "chr2"),
                  ranges = IRanges(start = c(150, 175, 550), 
                                   end = c(250, 450, 650),
                                   names=c("A","B","C")),
                  score = c(20, 40, 20),
                  strand="+")
```

<BR>

**Check the `GRanges` object**
```R
# check the GRanges object
query
subject
```

```text
GRanges object with 3 ranges and 1 metadata column:
    seqnames    ranges strand |     score
       <Rle> <IRanges>  <Rle> | <numeric>
  a     chr1   100-200      + |        10
  b     chr1   300-400      + |        20
  c     chr1   500-600      + |        30
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths
GRanges object with 3 ranges and 1 metadata column:
    seqnames    ranges strand |     score
       <Rle> <IRanges>  <Rle> | <numeric>
  A     chr1   150-250      + |        20
  B     chr1   175-450      + |        40
  C     chr2   550-650      + |        20
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths
```

<BR>

You can use **bed file** as input to creat `GRanges` objects
```R
peak1 <- toGRanges("peak1.bed", format="BED")
peak2 <- toGRanges("peak2.bed", format="BED")
```

<BR>

### IRanges objects

- `IRanges`(IntegerRanges) **lack** the information of `seqnames` and `strand` and add the width information comparing to `GRanges`

```R
# åˆ›å»º IRanges å¯¹è±¡
ir <- IRanges(start = c(1, 3, 20), end = c(13, 7, 25))

# æŸ¥çœ‹å¯¹è±¡
ir
```

```text
IRanges object with 3 ranges and 0 metadata columns:
          start       end     width
      <integer> <integer> <integer>
  [1]         1        13        13
  [2]         3         7         5
  [3]        20        25         6
```

<br>

***
## NCList and GNCList objects

- å¦‚æœqueryä¸­æœ‰$Q$ä¸ªåŒºé—´ï¼Œsubjectä¸­æœ‰$S$ä¸ªåŒºé—´ï¼Œä¼ ç»Ÿæ–¹æ³•éœ€è¦è¿›è¡Œ$QÃ—S$æ¬¡æ¯”è¾ƒã€‚
    <br>
- `NCList` é€šè¿‡åˆ†åŒºå’Œæ’åºï¼ŒåµŒå¥—åˆ†ç»„ï¼Œè¿™æ ·ï¼Œæ¯æ¬¡æœç´¢æ—¶åªéœ€è¦åœ¨ç›¸å…³å±‚æ¬¡ä¸­å¯»æ‰¾ï¼Œè€Œä¸éœ€è¦æ‰«ææ‰€æœ‰åŒºé—´

	  <img src="./picture/005.png" width="700"/>
	<BR>
- `GNCList`æ˜¯`NCList`çš„æ‰©å±•ç‰ˆæœ¬ï¼Œä¸“é—¨ç”¨äºå¤„ç†å¸¦æœ‰åŸºå› ç»„åºåˆ—åç§°ï¼ˆ**seqnames**ï¼‰çš„èŒƒå›´å¯¹è±¡ã€‚

	| ç‰¹æ€§       | `NCList`       | `GNCList`                    |
	| -------- | -------------- | ---------------------------- |
	| **é€‚ç”¨èŒƒå›´** | æ— æŸ“è‰²ä½“ä¿¡æ¯ï¼Œä»…å¤„ç†ä¸€èˆ¬èŒƒå›´ | é€‚ç”¨äºå¸¦æœ‰æŸ“è‰²ä½“ï¼ˆ`seqnames`ï¼‰ä¿¡æ¯çš„åŸºå› ç»„èŒƒå›´ |
	| **åˆ†ç»„**   | æ— éœ€åˆ†ç»„ï¼Œä»…ä½¿ç”¨åµŒå¥—ç»“æ„   | æŒ‰æŸ“è‰²ä½“åˆ†ç»„ï¼Œæ¯ä¸ªç»„å†…ä½¿ç”¨ NCList         |
	| **æ€§èƒ½**   | é€‚ç”¨äºå°å‹èŒƒå›´æ•°æ®      | é’ˆå¯¹å¤§å‹åŸºå› ç»„æ•°æ®ä¼˜åŒ–ï¼Œé¿å…è·¨æŸ“è‰²ä½“æ¯”è¾ƒ         |

<BR>

- `findOverlaps()` å’Œ `countOverlaps()`ä¼šè‡ªåŠ¨å¯¹ subject åˆ›å»º `NCList`å’Œ`GNCList`ï¼Œä»¥ä¼˜åŒ–é‡å è®¡ç®—çš„æ•ˆç‡ã€‚ç”¨æˆ·é€šå¸¸æ— éœ€æ‰‹åŠ¨æ„å»º `NCList`å’Œ`GNCList`ã€‚
  

```R
# creat nclist and gnclist

# å½“ä½ å¤„ç†åƒç¯çŠ¶æŸ“è‰²ä½“ï¼ˆå¦‚çº¿ç²’ä½“ DNAï¼‰æ•°æ®æ—¶
# circle.lengthå®šä¹‰äº†â€œç¯é•¿
# å½“åŒºé—´è¶…å‡ºè¾¹ç•Œæ—¶ï¼Œä½¿ç”¨ circle.length å¯ä»¥è®©åŒºé—´â€œå›åˆ°èµ·ç‚¹â€ï¼Œæ¨¡æ‹Ÿç¯çŠ¶æ•°æ®
nclist <- NCList(ir,
                 #circle.length=NA_integer_
                   )

gnclist <- GNCList(subject)

nclist
gnclist
```

```text
NCList object with 3 ranges and 0 metadata columns:
          start       end     width
      <integer> <integer> <integer>
  [1]         1        13        13
  [2]         3         7         5
  [3]        20        25         6
GNCList object with 3 ranges and 1 metadata column:
    seqnames    ranges strand |     score
       <Rle> <IRanges>  <Rle> | <numeric>
  A     chr1   150-250      + |        20
  B     chr1   175-450      + |        40
  C     chr2   550-650      + |        20
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths
```

<br>

---
## TxDb and EnsDb

### TxDb object

- `TxDb` (Transcript Database) is a object provided by Bioconductor for storing annotation information of genomes
    <br>
- you can download from `Bioconductor`

	```R
	# loading packages
	library(AnnotationHub)
	
	# downland AnnotationHub database
	AH <- AnnotationHub()
	
	# Query all available Txdb databases
	query(AH , "Txdb.Rnorvegicus.UCSC.rn6")
	```
	
	```text
	AnnotationHub with 6 records
	# snapshotDate(): 2024-10-28
	# $dataprovider: UCSC
	# $species: Rattus norvegicus
	# $rdataclass: TxDb
	# additional mcols(): taxonomyid, genome, description, coordinate_1_based, maintainer,
	#   rdatadateadded, preparerclass, tags, rdatapath, sourceurl, sourcetype 
	# retrieve records with, e.g., 'object[["AH52270"]]' 
	
	            title                                      
	  AH52270 | TxDb.Rnorvegicus.UCSC.rn6.refGene.sqlite   
	  AH53781 | TxDb.Rnorvegicus.UCSC.rn6.refGene.sqlite   
	  AH61799 | TxDb.Rnorvegicus.UCSC.rn6.refGene.sqlite   
	  AH66179 | TxDb.Rnorvegicus.UCSC.rn6.refGene.sqlite   
	  AH70598 | TxDb.Rnorvegicus.UCSC.rn6.refGene.sqlite   
	  AH84143 | TxDb.Rnorvegicus.UCSC.rn6.ncbiRefSeq.sqlite
	```
	
	According to **AH ID** to downland database
	```R
	# According to AH ID to downland database
	# A larger AH ID indicates a newer version.
	TxDb <- AH[["AH70598"]]
	```

  <br>
* you can use `GFF/GTF` to build `TxDb` object  

	<img src="./picture/013.jpg" width="700"/>
	```R
	# loading packages
	library(GenomicFeatures)
	
	# makeTxDb
	TxDb <- makeTxDbFromGFF("./Rattus_norvegicus.mRatBN7.2.113.gtf")	
	```

<BR>

### EnsDb

- æˆ‘ä»¬è¿˜å¯ä»¥ä½¿ç”¨ **Ensembl-based `EnsDb` æ³¨é‡Šæ•°æ®åº“ï¼ˆç”± [ensembldb](http://127.0.0.1:28103/session/Rvig.861c7c7c45e9.html) åŒ…æä¾›ï¼‰ä½œä¸º`TxDb` objects è¿›è¡Œè¾“å…¥

- ä½†éœ€è¦ç”¨`seqlevelsStyle()` å¯¹æŸ“è‰²ä½“åç§°è¿›è¡Œè½¬æ¢

|ç‰¹æ€§|`EnsDb`|`TxDb`|
|---|---|---|
|**æ•°æ®æ¥æº**|Ensemblæ•°æ®åº“çš„åŸºå› ç»„æ³¨é‡Šæ•°æ®|å¤šç§æ³¨é‡Šæºï¼ˆå¦‚UCSCã€NCBIï¼‰|
|**æ•°æ®å†…å®¹**|åŸºå› ã€è½¬å½•æœ¬ã€å¤–æ˜¾å­ã€CDSç­‰å…¨é¢ä¿¡æ¯|ä¸»è¦åŒ…å«è½¬å½•æœ¬ä¿¡æ¯ï¼Œå¦‚å¤–æ˜¾å­ã€å†…å«å­|
|**é€‚ç”¨åœºæ™¯**|åŸºå› ç»„çº§åˆ«çš„åˆ†æï¼Œç‰¹åˆ«æ˜¯Ensemblæ³¨é‡Š|è½¬å½•æœ¬åˆ†æï¼Œå…³æ³¨è½¬å½•æœ¬ç»“æ„ä¸åŠŸèƒ½|
```R
# loading packages
library(ensembldb)

# downland AnnotationHub database
AH <- AnnotationHub()

# Query all available Txdb databases
# This 113 is the version number for the EnsDb database
query(AH, "113 EnsDb for Rattus norvegicus")

# According to AH ID to downland database
edb <- AH[["AH119437"]]

# Change the chromosome names to UCSC style
seqlevelsStyle(edb) <- "UCSC"
```
<br>

---
## GFF(GFF3) and GTF(GTF2)

**[åŸºå› ç»„æ³¨é‡Šæ–‡ä»¶(GFF,GTF)ä¸‹è½½çš„äº”ç§æ–¹æ³•](https://blog.csdn.net/u011262253/article/details/89363809)**


### GFF(GFF3)

- `GFF`(General Feature Format)æ˜¯ä¸€ç§ç”¨æ¥æè¿°åŸºå› ç»„ç‰¹å¾çš„æ–‡ä»¶
  
- ç°åœ¨æˆ‘ä»¬æ‰€ä½¿ç”¨çš„å¤§éƒ¨åˆ†éƒ½æ˜¯ç¬¬ä¸‰ç‰ˆï¼ˆ`[GFF3](https://gmod.org/wiki/GFF3)`ï¼‰
  
    ```text
    ctg123 . mRNA            1300  9000  .  +  .  ID=mrna0001;Name=sonichedgehog
    ctg123 . exon            1300  1500  .  +  .  ID=exon00001;Parent=mrna0001
    ctg123 . exon            1050  1500  .  +  .  ID=exon00002;Parent=mrna0001
    ctg123 . exon            3000  3902  .  +  .  ID=exon00003;Parent=mrna0001
    ctg123 . exon            5000  5500  .  +  .  ID=exon00004;Parent=mrna0001
    ctg123 . exon            7000  9000  .  +  .  ID=exon00005;Parent=mrna0001
    ```
    
	|åˆ—å|æè¿°|
	|---|---|
	|seqid|å‚è€ƒåºåˆ—çš„ ID|
	|source|è¯¥ç‰¹å¾çš„æ¥æºï¼ŒæœªçŸ¥æ—¶ç”¨ `.` ä»£æ›¿|
	|type|ç‰¹å¾ç±»å‹ï¼Œå¦‚ `gene`ã€`exon`ã€`CDS` ç­‰ï¼Œå»ºè®®ä½¿ç”¨ SO æœ¯è¯­|
	|start|èµ·å§‹åæ ‡ï¼ˆ1-basedï¼‰|
	|end|ç»ˆæ­¢åæ ‡|
	|score|é‡åŒ–å¾—åˆ†ï¼Œç©ºå€¼ç”¨ `.` ä»£æ›¿|
	|strand|`+` è¡¨ç¤ºæ­£é“¾ï¼Œ`-` è¡¨ç¤ºè´Ÿé“¾ï¼Œ`.` è¡¨ç¤ºæ— æ–¹å‘è¦æ±‚|
	|phase|CDS éœ€è¦æ­¤åˆ—ï¼Œå½“å‰ CDS ç‰‡æ®µçš„èµ·å§‹ç¢±åŸºåœ¨å¯†ç å­ä¸­çš„åç§»é‡ã€‚å®ƒçš„å€¼å¯ä»¥æ˜¯ 0ã€1 æˆ– 2|
	|attributes|å±æ€§åˆ—è¡¨ï¼Œæ ¼å¼ä¸º `key=value`ï¼Œä¸åŒå±æ€§ç”¨ `;` åˆ†éš”|
    

<br>

### GTF (GTF2)

- `GTF`(gene transfer format)ï¼Œä¸»è¦æ˜¯ç”¨æ¥å¯¹åŸºå› è¿›è¡Œæ³¨é‡Š
  
- å½“å‰æ‰€å¹¿æ³›ä½¿ç”¨çš„GTFæ ¼å¼ä¸ºç¬¬äºŒç‰ˆï¼ˆ`GTF2`ï¼‰
  
    ```text
    AB000381 Twinscan  CDS          700   707   .   +   2  gene_id "AB000381.000"; transcript_id "AB000381.000.1";
    AB000381 Twinscan  exon         900  1000   .   +   .  gene_id "AB000381.000"; transcript_id "AB000381.000.1";
    AB000381 Twinscan  start_codon  380   382   .   +   0  gene_id "AB000381.000"; transcript_id "AB000381.000.1";
    AB000381 Twinscan  stop_codon   708   710   .   +   0  gene_id "AB000381.000"; transcript_id "AB000381.000.1";
    ```
    
	|åˆ—å|æè¿°|
	|---|---|
	|seqname|åºåˆ— IDï¼Œé€šå¸¸æ˜¯æŸ“è‰²ä½“ ID æˆ– Contig ID|
	|source|è¯¥ç‰¹å¾çš„æ¥æºï¼ŒæœªçŸ¥æ—¶ç”¨ `.` ä»£æ›¿|
	|type|ç‰¹å¾ç±»å‹ï¼Œç‰¹å¾ç±»å‹ï¼Œå¦‚ `CDS`ã€`start_codon`(èµ·å§‹å¯†ç å­)ã€`stop_codon`ï¼ˆç»ˆæ­¢å¯†ç å­ï¼‰ ï¼ˆ**è¿™ä¸€åˆ—å¿…é¡»æ³¨æ˜**ï¼‰|
	|start|èµ·å§‹åæ ‡ï¼ˆ1-basedï¼‰|
	|end|ç»ˆæ­¢åæ ‡|
	|score|é‡åŒ–å¾—åˆ†ï¼Œç©ºå€¼ç”¨ `.` ä»£æ›¿|
	|strand|`+` è¡¨ç¤ºæ­£é“¾ï¼Œ`-` è¡¨ç¤ºè´Ÿé“¾ï¼Œ`.` è¡¨ç¤ºæ— æ–¹å‘è¦æ±‚|
	|frame|è¯»ç æ¡†ï¼ˆ0ã€1 æˆ– 2ï¼Œä»…é€‚ç”¨äº `CDS`ï¼‰|
	|attributes|å…¶ä»–ä¿¡æ¯ï¼ˆé”®å€¼å¯¹ï¼Œ`key "value";` æ ¼å¼ï¼‰|

<br>

### Different from GFF of GTF

| å¯¹æ¯”é¡¹           | `GFF3`                                         | `GTF2`                             |
| ------------- | -------------------------------------------------- | -------------------------------------- |
| type/feature  | å¿…é¡»æ³¨æ˜                                               | å¯ä»¥æ˜¯ä»»æ„åç§°                                |
| attributes    | `key=value`                                        | `key "value"`                          |
| attributes å±‚çº§ | æ˜ç¡®å±‚çº§å…³ç³»ï¼ˆgene â†’ mRNA â†’ exon/CDSï¼‰                     | å±‚çº§å…³ç³»ç›¸å¯¹ä¸ä¸¥æ ¼                              |
| æ”¯æŒçš„ç‰¹å¾         | æ›´ä¸°å¯Œï¼ˆgene, mRNA, exon, CDS, ncRNA, repeat_region ç­‰ï¼‰ | ä¸»è¦é’ˆå¯¹è½¬å½•ç»„æ•°æ®ï¼ˆgene, transcript, exon, CDSï¼‰ |
<br>

---
# Genome Coordinate Conversion

## Loading package
```R
# Delete all objects
rm(list = ls())

# Loading package
library(R.utils)      # enhance the basic functions of R.
library(rtracklayer)  # processing genomic data.
library(dplyr)        # Data Processing and Operations
```

<BR>

---

## Processing data

Processing bed file as `GRanges` (`GRanges` objects as input in `liftOver()`

- you can import bed file as dataframe to create `GRanges` object**(Recommend)**ã€‚
  
- You should modify the chromosome names to conform to the genome format (e.g., **1** should become **chr1**)
    **UCSC** use **chr1** and **chrM** and version
    
	```R
	    # import bed file as dataframe
	bed_data <- read.csv(file = "./Demo0625.csv")
	
	# remove duplicate
	bed_data <- bed_data %>% distinct(bed_data$name,.keep_all = TRUE)
	
	
	# create GRanges objects
	query <- GRanges(seqnames = bed_data$chr,
	                  ranges = IRanges(start = bed_data$start, 
	                                   end = bed_data$end,
	                                   names=bed_data$name),
	                  score = bed_data$fold_enrichment,
	                  strand="*")
	
	# ä¿®æ”¹æŸ“è‰²ä½“åç§°
	seqlevelsStyle(query) <- "UCSC" 
	
	
	query
	```

	```text
	GRanges object with 10714 ranges and 1 metadata column:
	                    seqnames              ranges strand |     score
	                       <Rle>           <IRanges>  <Rle> | <numeric>
	      KChIP2_peak_3     chr1     1193628-1193990      * |   3.49474
	      KChIP2_peak_4     chr1     1670780-1671341      * |   6.46071
	      KChIP2_peak_6     chr1     1784527-1784797      * |   4.72403
	      KChIP2_peak_7     chr1     1803427-1803932      * |   3.62225
	      KChIP2_peak_8     chr1     1925707-1926101      * |   2.62105
	                ...      ...                 ...    ... .       ...
	  KChIP2_peak_21434     chrX 158978724-158979272      * |   4.24629
	  KChIP2_peak_21436     chrX 159126572-159127036      * |   3.43400
	  KChIP2_peak_21437     chrX 159157346-159157724      * |   3.72489
	  KChIP2_peak_21438     chrX 159428723-159429014      * |   2.62105
	  KChIP2_peak_21441     chrX 159890957-159891282      * |   2.62105
	  -------
	  seqinfo: 22 sequences from an unspecified genome; no seqlengths
	```

-  you can use `import()` function to directly import bed file as GRange object**(NOT recommend)**ã€‚
	```R
	query <- import("./Demodata.bed")
	```

<BR>

---
## Chain file

`Chain file` is a file format used to represent **sequence alignments between different genome assemblies** of the same species or between different species.

<BR>

### Loading chain file

- **[UCSC](https://hgdownload.soe.ucsc.edu/downloads.html)** [chain file download web](https://hgdownload.soe.ucsc.edu/downloads.html) **(Recommend)**
    **(choose species ------choose genome version-----choose LifeOver files)**
    <BR>
- **[Ensembl](https://sourceforge.net/projects/crossmap/files/Ensembl_chain_files/)** [chain file download web](https://sourceforge.net/projects/crossmap/files/Ensembl_chain_files/) **(Only have human and mouse)**

```R
# decompressing chain file
if (file.exists("rn6ToRn7.over.chain")) {
  print("æ–‡ä»¶è§£å‹æˆåŠŸï¼")
} else {
  gunzip("rn6ToRn7.over.chain.gz")
  print("æ–‡ä»¶é‡æ–°è§£å‹å®Œæˆã€‚")
}


# read chain file
chain_file <- import.chain("rn6ToRn7.over.chain")
```

<BR>

### Structure of chain file

[Reference](https://genome.ucsc.edu/goldenpath/help/chain.html)

A `chain file` is generated by **alignment** two genomes

* `chain file`é‡Œé¢åŒ…å«è®¸å¤šå—**chain(åŒæºçš„åœ°æ–¹)**

	- each **chain** represents an alignment between two genomic regions
	    <br>
	- å…¶ä¸­æ¯å—**chain** ç”±**Header Line** å’Œ **Alignment Data Lines**ç»„æˆ
	    <br>
	- ç¬¬ä¸€è¡Œ`Header Line`è®°å½•**chain**åœ¨ä¸¤ä¸ªç‰ˆæœ¬ä¸­åæ ‡ï¼Œ
	    <br>
	- åé¢è®¸å¤šè¡Œ`alignment data line`è®°å½•å…·ä½“æ¯”å¯¹æƒ…å†µã€‚
	```text
	chain 4900 chrY 58368225 + 25985403 25985638 chr5 151006098 - 43257292 43257528 1
	9 1 0
	10 0 5
	61 4 0
	16 0 4
	42 3 0
	16 0 8
	14 1 0
	3 7 0
	48
	
	chain 4900 chrY 58368225 + 25985406 25985566 chr5 151006098 - 43549808 43549970 2
	16 0 2
	60 4 0
	10 0 4
	```

<br>

#### Header Line
```text
chain score tName tSize tStrand tStart tEnd qName qSize qStrand qStart qEnd id 
```

- `score` -- chain score ï¼ˆHow well the two sequences allignï¼‰
  
- `tName` -- chromosome (reference/target sequence)
  
- `tSize` -- chromosome size (reference/target sequence)
  
- `tStrand` -- strand (reference/target sequence)
  
- `tStart` -- alignment start position (reference/target sequence)
  
- `tEnd` -- alignment end position (reference/target sequence)
  
- `qName` -- chromosome (query sequence)
  
- `qSize` -- chromosome size (query sequence)
  
- `qStrand` -- strand (query sequence)
  
- `qStart` -- alignment start position (query sequence)
  
- `qEnd` -- alignment end position (query sequence)
  
- `id` -- chain ID

<br>

#### Alignment Data Lines
```text
size dt dq 
```
- Alignment Data Lines å°† chainåˆ†æˆè®¸å¤šä¸ª **block**(ä¸€è‡´çš„åŒºåŸŸ)(å¯¹é½å—)ï¼Œæ¯ä¸€è¡Œä»£è¡¨ä¸€ä¸ª**block**
  
- `size` -- the size of the ungapped alignment **(block)**
  
- `dt` -- the difference between the end of this **block** and the beginning of the next **block** **(reference/target sequence)**
  
- `dq` -- the difference between the end of this **block** and the beginning of the next **block** **(query sequence)**
```text
     chain 4900 chrY 58368225 + 25985406 25985566 chr5 151006098 - 43549808 43549970 2
     16      0       2
     60      4       0
     10      0       4
     70 
```

<img src="./picture/001.png" width="800"/>
<br>

---
## Genome Coordinate Conversion

We can use function of `liftOver()` to transformGenome Coordinate Conversion in R
```r
# liftover
subject <- liftOver(query, chain_file)

# convert to dataframe
 subject_data <- as.data.frame(subject)

```

<br>

---
## Merge separate peaks

- Filter and merge peaks by setting `distance_threshold` and `minimum_width`
  
- è¿™ä¸ª`distance_threshold`ä¸`minimum_width`å€¼éœ€è¦è‡ªå·±æŠŠæ¡ï¼Œéœ€è¦æ ¹æ®å®é™…æƒ…å†µè¿›è¡Œåˆ¤æ–­**
  
- è¿™ä¸ªæ–¹æ³•æ˜¯æˆ‘ä¸€æ—¶åŠäº†æƒ³å‡ºæ¥ï¼Œå¦‚æœæœ‰å¥½çš„æƒ³æ³•å¯ä»¥è¿›è¡Œæ·»åŠ ä¸ä¿®æ”¹ã€‚
<img src="./picture/002.png" width="700"/>
<br>
<img src="./picture/003.png" width="700"/>
<br>
<img src="./picture/004.png" width="700"/>
<br>

### Per setting
```R
# setting distance_threshold (base)
distance_threshold <- 30

# setting minimum width of peaks (base)
minimum_width <- 30
```

<BR>

### Function of merge peaks by group
We made a `merge_peaks_by_group`function to merge peaks by same group
```R
merge_peaks_by_group <- function(df, distance_threshold) {
  df %>%
    group_by(group) %>%  # æŒ‰ group åˆ†ç»„
    filter(n_distinct(seqnames) == 1) %>%  # åªä¿ç•™åŒä¸€ç»„å†…æŸ“è‰²ä½“ç›¸åŒçš„è¡Œ
    arrange(group, start) %>%  # æŒ‰ç…§ start ä½ç½®æ’åº
    mutate(
      dist_to_prev = start - lag(end, default = first(end))  # å¢åŠ dist_to_prevè¿™åˆ—ï¼Œå…¶ä¸ºå½“å‰ start å’Œå‰ä¸€ä¸ª peak çš„ end ä¹‹é—´çš„è·ç¦»
    ) %>%
    ungroup() %>%  # å–æ¶ˆåˆ†ç»„ï¼Œä»¥ä¾¿ä¸‹ä¸€æ­¥æ“ä½œ
    mutate(
      merge_group = cumsum(dist_to_prev > distance_threshold | is.na(dist_to_prev))  # å¢åŠ merge_groupåˆ—ï¼Œå¦‚æœè·ç¦»å¤§äºé˜ˆå€¼ï¼Œå¢åŠ 1ï¼ˆä¸åˆå¹¶ï¼‰
    ) %>%
    group_by(group, merge_group) %>%  # æŒ‰ç…§ group å’Œæ–°çš„merge_groupè¿›è¡Œåˆ†ç»„
    summarise(
      group_name = first(group_name),  # ä¿ç•™ç¬¬ä¸€ä¸ª group_name
      seqnames = first(seqnames),
      start = min(start),  # å–æœ€å°çš„ start
      end = max(end),  # å–æœ€å¤§çš„ end
      width = end - start+1,  # è®¡ç®—æ–°çš„ width
      strand = first(strand),  # ä¿ç•™ç¬¬ä¸€ä¸ª strand
      score = first(score),  # ä¿ç•™ç¬¬ä¸€ä¸ª score
      .groups = "drop"  # å»æ‰åˆ†ç»„ç»“æ„
    )
}
```

<br>

Use `merge_peaks_by_group`function
```r
merge_peaks_subject <- merge_peaks_by_group(subject_data,distance_threshold)
```

### Remove narrow width of peaks

- Evaluate the change in width after `liftOver` to assess the quality of merged peaks
  
- peak whose width **increases too much** after `liftOver` is **removed**
```r
# Extract original width
before_width <- bed_data[,c(22,4)]
colnames(before_width) <- c("group_name","before_width")


# merge original width 
merge_peaks_subject <- merge(merge_peaks_subject,before_width,by.x="group_name",by.y="group_name")


# calculate the change in width after liftOver to assess the quality
merge_peaks_subject$width_change <- merge_peaks_subject$width/merge_peaks_subject$before_width*100


# remove narrow width of peaks
merge_peaks_subject <- filter(merge_peaks_subject,merge_peaks_subject$width>=minimum_width)
```

<br>

---
# Peak overlap

## Loading package
```r
# Delete all objects
rm(list = ls())

# Loading package
library(GenomicRanges)
library(ChIPpeakAnno)
```

- `GenomicRanges`æ˜¯Rè¯­è¨€ä¸­ä¸€ä¸ªå¼ºå¤§è€Œçµæ´»çš„ç”Ÿç‰©ä¿¡æ¯å­¦åŒ…ï¼Œä¸»è¦ç”¨äº**å¤„ç†åŸºå› ç»„åŒºé—´æ•°æ®**
  
- `ChIPpeakAnno` æ˜¯ä¸€ä¸ªä¸“æ³¨äº ChIP-seq æ•°æ®åˆ†æçš„ R åŒ…ï¼Œå¹¿æ³›ç”¨äºåˆ†æå’Œæ³¨é‡ŠæŸ“è‰²è´¨å…ç–«æ²‰æ·€ (ChIP) å®éªŒæ£€æµ‹çš„å³°æ•°æ®ã€‚å®ƒæä¾›äº†ä¸€ç³»åˆ—å¼ºå¤§çš„å·¥å…·ï¼Œç”¨äº**å³°é‡å åˆ†æã€åŸºå› æ³¨é‡Šã€åŠŸèƒ½å¯Œé›†åˆ†æå’Œå¯è§†åŒ–ç­‰**ã€‚

<br>

---
## Find the overlaps

- `findOverlaps()` å’Œ `countOverlaps()`ä¼šè‡ªåŠ¨å¯¹ subject åˆ›å»º **NCList**å’Œ**GNCList**ï¼Œä»¥ä¼˜åŒ–é‡å è®¡ç®—çš„æ•ˆç‡ã€‚ç”¨æˆ·é€šå¸¸æ— éœ€æ‰‹åŠ¨æ„å»º **NCList**å’Œ**GNCList**ã€‚

<br>

### findOverlaps()

We use `findOverlaps()` to find the overlaps

```r
# creat two GRanges object
query <- GRanges(seqnames = c("chr1","chr1","chr1"),
                  ranges = IRanges(start = c(100, 300, 500), 
                                   end = c(200, 400, 600),
                                   names=c("a","b","c")),
                  score = c(10, 20, 30),
                  strand="+")

subject <- GRanges(seqnames = c("chr1", "chr1", "chr2"),
                  ranges = IRanges(start = c(150, 175, 550), 
                                   end = c(250, 450, 650),
                                   names=c("A","B","C")),
                  score = c(20, 40, 20),
                  strand="+")
```

```r
hits <- findOverlaps(query, #  query GRanges object
                         subject, #  subject GRanges object
                         type="any",
                         #minoverlap=1, # ä¸¤ä¸ªåŒºåŸŸä¹‹é—´çš„æœ€å°é‡å ç¢±åŸºæ•°
                         #maxgap=0, # å…è®¸é‡å åŒºåŸŸä¹‹é—´çš„æœ€å¤§é—´éš™
                         ignore.strand= F, # F means strand ising considered
                         select="all",
                         # invert=TRUE #If True, keep only the ranges in x that do not overlap ranges
                         )
```

<br>

- `type`
  
    - `any`: è®¤ä¸ºqueryå’Œsubjectå­˜åœ¨ä»»æ„çš„é‡åˆå³è®¤ä¸ºå‘ç”Ÿäº†é‡å **ï¼ˆdeafultï¼‰**
      
    - `within`: è®¤ä¸ºqueryåªæœ‰å®Œå…¨åœ¨subjectå†…æ‰è®¤ä¸ºæ˜¯é‡å 
      
    - `start`: è®¤ä¸ºqueryä¸­çš„startä½ç½®åœ¨subjecttå†…æ‰è®¤ä¸ºæ˜¯é‡å 
      
    - `end`: è®¤ä¸ºqueryä¸­çš„endä½ç½®åœ¨subjectå†…æ‰è®¤ä¸ºæ˜¯é‡å 
      
    - `equal`ï¼šè®¤ä¸ºqueryä¸­çš„startï¼Œendä¸subjectå®Œå…¨ç›¸åŒæ‰è®¤ä¸ºæ˜¯é‡å 
      
    - When type is "`any`", at least one of maxgap and minoverlap must be set to its default value.
	<img src="./picture/006.png" width="700"/>

<br>

- `maxgap and minoverlap`
  
    - **Positive Values of maxgapï¼š**if `maxgap = 50`, two ranges can be up to 50 base pairs apart and still be considered overlapping
      
    - **Negative Values of maxgap**ï¼š if `maxgap = -20`, the two ranges must overlap by at least 20 base pairs.
      
    - `minoverlap`ï¼š**Only ranges with a minimum of `minoverlap` overlapping positions are considered to be overlapping.
	<img src="./picture/007.png" width="700"/>

<br>

- `select`
  
    - `all`: Return all overlaps **(default)**.
      
    - `first`: Return the first overlap for each **query range**
      
    - `last`: Return the last overlap for each **query range**
      
    - `arbitrary`: Return one arbitrary overlap for each **query range**
	<img src="./picture/008.png" width="700"/>

<br>

### hits objects

- The function of `findoverlaps()` function will creat `hits` ** objects**

	<img src="./picture/009.png" width="300"/>

```r
hits
```

```text
Hits object with 3 hits and 0 metadata columns:
      queryHits subjectHits
      <integer>   <integer>
  [1]         1           1
  [2]         1           2
  [3]         2           2
  -------
  queryLength: 3 / subjectLength: 3
```

<br>

- We can use `countQueryHits()`extracting numbers of overlap of query range**
  Using `countSubjectHits()` extracting numbers of overlap of subject range

	```r
	# Query
	countQueryHits(hits)
	setNames(countQueryHits(hits), seqnames(subject))
	
	# subject
	countSubjectHits(hits)
	setNames(countSubjectHits(hits), seqnames(subject))
	```
	
	```text
	[1] 2 1 0
	chr1 chr1 chr2 
	   2    1    0 
	[1] 1 2 0
	chr1 chr1 chr2 
	   1    2    0
	```
<br>

* Using `overlapsRanges()` to check the Specific range of overlaps
```r
GenomicRanges::intersect(query, subject)
```

```text
GRanges object with 2 ranges and 0 metadata columns:
      seqnames    ranges strand
         <Rle> <IRanges>  <Rle>
  [1]     chr1   150-200      +
  [2]     chr1   300-400      +
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths
```
<br>

### countOverlaps()

- `countOverlaps()`å¯ä»¥ç›´æ¥æŸ¥çœ‹queryä¸subjecté‡å çš„æ•°ç›®
  
- `countOverlaps()` **å ç”¨çš„å†…å­˜éå¸¸å°‘ï¼Œå†å¿«é€ŸæŸ¥çœ‹é‡å æ•°ç›®éå¸¸æœ‰ç”¨**
  
- `countOverlaps()` ç”¨æ³•ä¸`findOverlaps`ä¸€æ ·

```r
overlap_counts <- countOverlaps(query, subject)
overlap_counts
```

```text
a b c 
2 1 0
```

<br>

### Extract the overlapping peaks

The function of `pintersect()` find the overlapping portion of each range in one set with its paired range in other set

```r
# Extract overlapping ranges
overlap_ranges <- pintersect(query[queryHits(hits)], 
                             subject[subjectHits(hits)])

# caulating the mean score
mean_scores <- rowMeans(cbind(
  query$score[queryHits(hits)],
  subject$score[subjectHits(hits)]
))

# # Add the mean scores as metadata to the overlap ranges
overlap_ranges$score <- mean_scores

overlap_ranges
```

```text
GRanges object with 3 ranges and 2 metadata columns:
    seqnames    ranges strand |     score       hit
       <Rle> <IRanges>  <Rle> | <numeric> <logical>
  a     chr1   150-200      + |        15      TRUE
  a     chr1   175-200      + |        25      TRUE
  b     chr1   300-400      + |        30      TRUE
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths
```

<br>

### Find Overlaps with â‰¥50%(threshold)
```r
# setting the threshold
threshold=0.5


# exact the overlap_widths
overlap_widths <- width(overlap_ranges)


# Calculate widths of query and subject ranges
query_widths <- width(query[queryHits(hits)])
subject_widths <- width(subject[subjectHits(hits)])



# Filter overlaps with at least threshold overlap
keep <- (overlap_widths / query_widths >= threshold) | 
        (overlap_widths / subject_widths >= threshold)



# Subset the overlaps
threshold_hits <- hits[keep]


# Extract the filtered intersected ranges
threshold_overlap_ranges <- pintersect(query[queryHits(threshold_hits)], 
                                          subject[subjectHits(threshold_hits)])



# check the result
threshold_overlap_ranges
```

```text
GRanges object with 2 ranges and 2 metadata columns:
    seqnames    ranges strand |     score       hit
       <Rle> <IRanges>  <Rle> | <numeric> <logical>
  a     chr1   150-200      + |        10      TRUE
  b     chr1   300-400      + |        20      TRUE
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths
```

<br>

---
## findOverlapsOfPeaks()

- `findOverlapsOfPeaks()` æ˜¯ `ChIPpeakAnno` åŒ…æä¾›çš„ä¸€ä¸ªå‡½æ•°ï¼Œç”¨äºåœ¨å³°æ–‡ä»¶ä¸­æŸ¥æ‰¾å¤šä¸ªç»„é—´çš„é‡å åŒºåŸŸï¼Œç”¨æ³•ç±»ä¼¼äº`findOverlaps()`
  
- `findOverlapsOfPeaks()`ç›¸æ¯”`findOverlaps()` æ”¯æŒ**Veneå›¾å¯è§†åŒ–ï¼Œpeakæ³¨é‡Šï¼Œpeakå¯è§†åŒ–ç­‰åŠŸèƒ½**ã€‚
  
- `findOverlapsOfPeaks()` ä¼šç”Ÿæˆä¸€ä¸ª `overlappingPeaks` **list.**

```r
query <- GRanges(seqnames = c("chr1","chr1","chr1","chr1","chr1"),
                 ranges = IRanges(start = c(100, 300, 500, 100, 135), 
                                  end = c(200, 400, 600, 400, 334),
                                  names=c("a","b","c","d","e")),
                 score = c(10, 20, 30, 24, 32),
                 strand="+")

subject <- GRanges(seqnames = c("chr1", "chr1","chr1","chr1"),
                   ranges = IRanges(start = c(150, 175, 550,700), 
                                    end = c(250, 450, 650,750),
                                    names=c("A","B","C","D")),
                   score = c(20, 40, 20,30),
                   strand="+")
```

```r
overlappingPeaks <- findOverlapsOfPeaks(query,
                                     subject,
                                     maxgap=0, # å…è®¸é‡å åŒºåŸŸä¹‹é—´çš„æœ€å¤§é—´éš™
                                     minoverlap=0.5, # ä¸¤ä¸ªåŒºåŸŸä¹‹é—´çš„æœ€å°é‡å ç™¾åˆ†æ¯”
                                     ignore.strand= F, # F means strand ising considered
                                     connectedPeaks="keepAll")
```

* å€¼å¾—æ³¨æ„ï¼šè¿™é‡Œçš„`minoverlap`ä¸`findOverlaps()` ä¸åŒï¼š`findOverlaps()`æŒ‡æœ€å°é‡å base pairsã€‚è¿™é‡ŒæŒ‡çš„æ˜¯çš„æ˜¯overlapæœ€å°ç™¾åˆ†æ¯”(0 < `minoverlap` < 1)
<br>
- `connectedPeaks`:
  
    - if 5 peaks in query overlap with 2 peaks in subject
      
    - setting it to `merge` will add 1 to the `Counts`
      
    - setting it to `keepAll` will add 5 peaks to `count.query`, 2 to `count.subject`, and 2 to `Counts`
      
    - setting it to `min` will add 2 to the `Counts`
      
    
    è¿™é‡Œçš„`Counts`æ˜¯æŒ‡** `overlappingPeaks` **list ä¸­çš„** `venn_cnt` objectï¼Œç”¨äºéŸ¦æ©å›¾ç”»å›¾ç”¨
	```r
	overlappingPeaks[["venn_cnt"]]
	```

	```text
	     query subject Counts count.query count.subject
	[1,]     0       0      0           0             0
	[2,]     0       1      1           0             1
	[3,]     1       0      0           0             0
	[4,]     1       1      3           5             3
	attr(,"class")
	[1] "VennCounts"
	```

	<img src="./picture/010.png" width="600"/>
	
	`count.query`åœ¨ query ä¸­æœ‰å¤šå°‘ä¸ªpeaksä¸ subject ä¸­çš„peakså‘ç”Ÿäº†é‡å ã€‚
	`count.subject`åœ¨ subject ä¸­æœ‰å¤šå°‘ä¸ªpeaksä¸ query ä¸­çš„peakså‘ç”Ÿäº†é‡å ã€‚
	`Counts`åœ¨ `keepAll` ä¸ `min` æŒ‡`count.query`ä¸`count.subject`æœ€å°å€¼ï¼Œåœ¨ `merge` æŒ‡æ˜¯å°†é‡å çš„peaksåˆå¹¶ä¸ºä¸€ä¸ªå•ä¸€è®¡æ•°çš„ç»“æœã€‚
	<img src="./picture/011.png" width="700"/>

<br>

---
## overlappingPeaks list

`overlappingPeaks` list ç”±ä»¥ä¸‹objectsæ„æˆ

- `venn_cnt` :** an **object** of `VennCounts` ï¼Œç”¨äºéŸ¦æ©å›¾ç”»å›¾ç”¨
  
- `peaklist`ï¼š** a **list** of `GRanges` object which contain all **overlapping peaks** or **unique peaks**
  
- `uniquePeaks`ï¼š** an **object** of `GRanges` consists of all **unique peaks**
  
- `mergedPeaks`ï¼š** an **object** of `GRanges` consists of all **merged overlapping peaks**
  
- `peaksInMergedPeaks`ï¼š** an **object** of `GRanges` consists of all peaks in each samples involved in the overlapping peaks
  
- `overlappingPeaks`ï¼š** a **list** of **data frame** consists of the annotation of all the overlapped peaks
  
- **all.peaksï¼š** a **list** of `GRanges` object which contain the **input peaks** with **formated rownames**.

<img src="./picture/012.png" width="700"/>

<br>

---
## Draw a vennDiagram

- `makeVennDiagram()` from packages of `ChIPpeakAnno`
  
- `makeVennDiagram()` :Making Venn Diagram from two or more peak ranges, Also calculate p-value to determine whether those peaks overlap significantly.
```r
# per setting
# color
col = rainbow(2)
```

```r
VennDiagram <- makeVennDiagram(
               Peaks = list(query, subject),
               NameOfPeaks = c("query", "subject"),
               maxgap = 0L, # å…è®¸é‡å åŒºåŸŸä¹‹é—´çš„æœ€å¤§é—´éš™
               minoverlap = 0.5, # ä¸¤ä¸ªåŒºåŸŸä¹‹é—´çš„æœ€å°é‡å ç™¾åˆ†æ¯”
               totalTest = 1000, # è®¾ç½®æ€»æµ‹è¯•æ¬¡æ•°ï¼Œç”¨äºè®¡ç®— p å€¼ï¼ŒIt should be much larger than the number of peaks in the largest peak set.
               by = "region", 
               ignore.strand = T,
               connectedPeaks = "merge",
               method = "hyperG",
               # TxDb = TxDb #  TxDb objects   
               plot = TRUE,  # a venn diagram is plotted
               
               # venn plot setting
               alpha= 0.3, # é€æ˜åº¦
               scaled = T, # æ ¹æ®æ¯”ä¾‹æ˜¾ç¤ºå¤§å°
               
               lwd=1, # åœ†åœˆçº¿æ¡ç²—ç»†
               lty=1, # åœ†åœˆçº¿æ¡å½¢çŠ¶: 1 å®çº¿, 2 è™šçº¿
               col=col, #åœ†åœˆçº¿æ¡é¢œè‰²: blankæ— çº¿æ¡
               fill= col, # fill color
               
               # numeric labels
               cex = 1,  # Size
               label.col ='black', # color
               fontface = "bold", # å­—ä½“ç²—ç»†ï¼›åŠ ç²—bold

               # category labels
               cat.cex = 1, # Size
               cat.dist = 0.02, # æ ‡ç­¾è·ç¦»åœ†åœˆçš„è¿œè¿‘
               cat.pos = -180, # æ ‡ç­¾ç›¸å¯¹äºåœ†åœˆçš„è§’åº¦cat.pos = c(-10, 10, 135)
               cat.fontface = "bold",  # æ ‡ç­¾å­—ä½“åŠ ç²—
               cat.default.pos = "text"  # æ ‡ç­¾ä½ç½®, outerå†…;text å¤–
)
```

<img src="./picture/014.jpg" width="500"/>

<br>

- `by`
  
    * `feature` : means using feature field (such as Promoter, exon) for calculating overlap
    
    * `region`(default): means using chromosome range for calculating overlap,
    
    * `base`: means calculating overlap in nucleotide level.
    <br>
- `connectedPeaks`
	- è¿™é‡Œçš„`connectedPeaks`å¤šäº†ä¸€ä¸ªå‚æ•°ï¼š
    
    `keepFirstListConsistent` will keep the counts consistent with first list.
    <br>
- `method`ï¼šmethod to be used for p value calculation
  
    - `hyperG`ï¼ˆè¶…å‡ ä½•æ£€éªŒï¼‰ï¼šè¶…å‡ ä½•åˆ†å¸ƒä¼šè®¡ç®—å‡ºéšæœºæƒ…å†µä¸‹æŠ½ä¸­æŸä¸ªé‡å çš„æ¦‚ç‡ã€‚
      
    - `permutation`ï¼ˆç½®æ¢æ£€éªŒï¼‰ï¼šä½ å°†è¿™äº›æ•°æ®æ‰“ä¹±å¤šæ¬¡ï¼ˆå¦‚ï¼šå°† query å’Œ subject ä¸­çš„å³°éšæœºåˆ†é…åˆ°æ•´ä¸ª 1000ä¸ªåŒºé—´ï¼‰ï¼Œæ¯æ¬¡éƒ½è®¡ç®—ä¸€æ¬¡é‡å æƒ…å†µã€‚ç„¶åï¼Œä½ æ¯”è¾ƒåŸå§‹æ•°æ®çš„é‡å å’Œè¿™äº›éšæœºæ‰“ä¹±åçš„é‡å ç»“æœã€‚å¦‚æœåŸå§‹æ•°æ®çš„é‡å ç»“æœæ¯”æ‰“ä¹±åçš„ç»“æœæ˜¾è‘—ï¼ˆå³æ›´æç«¯æˆ–æ›´å¤§ï¼‰ï¼Œé‚£ä¹ˆè¯´æ˜é‡å ç»“æœæ˜¯æ˜¾è‘—çš„ï¼Œä¸å¤ªå¯èƒ½æ˜¯å¶ç„¶çš„ã€‚

<br>

---
# peak Annotation, Comparison and Visualization
## Loading packages
```r
# Delete all objects
rm(list = ls())

# Loading packages
library(ChIPseeker)      # for ChIP peak Annotation, Comparison and Visualization
library(dplyr)           # Data Processing and Operations
library(rtracklayer)     # processing genomic data
library(ggplot2)         # Data Visualization
library(org.Rn.eg.db)    # wide annotation database of Genome,choise your species
library(AnnotationHub)   # Query the annotation database
library(clusterProfiler) # enrichment analysis

library(knitr)      # for creating tables in RMarkdown
library(kableExtra) # for enhancing table styling
library(reactable)  # beautiful table
```

<br>

---

## Date processing
```r
# å®šä¹‰BEDæ–‡ä»¶è·¯å¾„å’Œpeak name
bed_files <- c("./Demo0625.csv", "./Demo1011.csv")
peak_name <- c("peak0625", "peak1011")

# åˆå§‹åŒ–ä¸€ä¸ªç©ºåˆ—è¡¨ï¼Œç”¨äºå­˜å‚¨æ¯ä¸ªæ–‡ä»¶çš„GRangeså¯¹è±¡
peak_list <- list()

# å¾ªç¯åŠ è½½æ¯ä¸ªBEDæ–‡ä»¶å¹¶è½¬æ¢ä¸ºGRangeså¯¹è±¡
for (i in seq_along(bed_files)) {
  
  # è¯»å–BEDæ–‡ä»¶
  bed_data <- read.csv(file = bed_files[i])
  
  # å»é™¤é‡å¤è¡Œ
  bed_data <- bed_data %>% distinct(name, .keep_all = TRUE)
  
  # åˆ›å»ºGRangeså¯¹è±¡
  peak <- GRanges(
    seqnames = bed_data$chr,
    ranges = IRanges(
      start = bed_data$start,
      end = bed_data$end,
      names = bed_data$name
    ),
    score = bed_data$fold_enrichment,
    strand = "*"
  )
  
  # ä¿®æ”¹æŸ“è‰²ä½“åç§°
  seqlevelsStyle(peak) <- "UCSC" 
  
  # å°†GRangeså¯¹è±¡æ·»åŠ åˆ°åˆ—è¡¨ä¸­ï¼Œå¹¶ä½¿ç”¨å¯¹åº”çš„peak_nameå‘½å
  peak_list[[peak_name[i]]] <- peak
  
  rm(peak)
  rm(bed_data)
}

# æŸ¥çœ‹ç»“æœ
peak_list
```

```text
$peak0625
GRanges object with 10714 ranges and 1 metadata column:
                    seqnames              ranges strand |     score
                       <Rle>           <IRanges>  <Rle> | <numeric>
      KChIP2_peak_3     chr1     1193628-1193990      * |   3.49474
      KChIP2_peak_4     chr1     1670780-1671341      * |   6.46071
      KChIP2_peak_6     chr1     1784527-1784797      * |   4.72403
      KChIP2_peak_7     chr1     1803427-1803932      * |   3.62225
      KChIP2_peak_8     chr1     1925707-1926101      * |   2.62105
                ...      ...                 ...    ... .       ...
  KChIP2_peak_21434     chrX 158978724-158979272      * |   4.24629
  KChIP2_peak_21436     chrX 159126572-159127036      * |   3.43400
  KChIP2_peak_21437     chrX 159157346-159157724      * |   3.72489
  KChIP2_peak_21438     chrX 159428723-159429014      * |   2.62105
  KChIP2_peak_21441     chrX 159890957-159891282      * |   2.62105
  -------
  seqinfo: 22 sequences from an unspecified genome; no seqlengths

$peak1011
GRanges object with 10019 ranges and 1 metadata column:
                       seqnames              ranges strand |     score
                          <Rle>           <IRanges>  <Rle> | <numeric>
      KChIP2-OE_peak_2     chr1       747955-748316      * |   2.84924
      KChIP2-OE_peak_5     chr1     1671192-1671604      * |   3.73480
      KChIP2-OE_peak_6     chr1     1767207-1767799      * |   4.68312
      KChIP2-OE_peak_7     chr1     1944415-1944899      * |   5.71260
      KChIP2-OE_peak_8     chr1     1952088-1952383      * |   2.84924
                   ...      ...                 ...    ... .       ...
  KChIP2-OE_peak_22997     chrX 159037695-159038091      * |   4.70768
  KChIP2-OE_peak_22999     chrX 159437682-159437965      * |   3.56155
  KChIP2-OE_peak_23001     chrX 159816201-159816424      * |   3.56155
  KChIP2-OE_peak_23002     chrX 159890739-159891449      * |   4.27386
  KChIP2-OE_peak_23003     chrY       525379-525712      * |   2.84924
  -------
  seqinfo: 23 sequences from an unspecified genome; no seqlengths
```

***
## Loading Txdb

`Txdb` detail you can see  ğŸª¨[[#TxDb and EnsDb]] ğŸ”—[TxDb and EnsDb](#TxDb-and-EnsDb)

```r
# downland AnnotationHub database
AH <- AnnotationHub()

TxDb <- AH[["AH70598"]]
```


<br>

---
## Peaks coverage plot
- After peak calling, we would like to know the peak locations **over the whole genome**

	```r
	covplot(peak_list[[1]], weightCol="score")
	```
	<img src="./picture/015.jpg" width="600"/>

<br>

- we can check peak locations of **region** **of the chromosome**

	```r
	covplot(peak_list[[1]], weightCol="score", chrs=c("chr17", "chr18"), xlim=c(1, 5e7))
	```
	<img src="./picture/016.jpg" width="600"/>


<br>

* We also check peak locations **over the whole genome** with a **list** of `GRangs` objects

	```r
	covplot(peak_list, weightCol = "score", fill_color = c("red","blue")) +
	  theme(legend.position = "inside",
	        legend.position.inside = c(0.8,0.2))
	```
	<img src="./picture/018.jpg" width="600"/>

<br>

---
## Upstream and downstream

- we only set `Upstream and downstream` to `500` as an example for only demonstration.User should set according to the actual situation(Deafut: `1000` or `3000`)
    <br>
- `Upstream and downstream` can be set using `rel(0.2)`, which represents an expansion of 20% of the target region's length rather than a fixed bp value. However, when `type = "body"`, using `rel()` will not cause any errors.

<br>

---
## nbin

|åˆ†æåœºæ™¯|`nbin` å»ºè®®å€¼|è¯´æ˜|
|---|---|---|
|TSS/TES Â±2kb åŒºåŸŸ|`100` - `400`|é€‚ç”¨äºå°åŒºåŸŸï¼Œ`nbin` é€‚ä¸­å¯ä¿è¯å¹³æ»‘æ›²çº¿|
|æ•´ä¸ªåŸºå› ä½“ (`type = "body"`)|`400` - `1000`|é€‚ç”¨äºå¤§èŒƒå›´åˆ†æ|
|è¶…å¤§åŸºå› åŒºåŸŸï¼ˆ>100kbï¼‰|>`1000`|é€‚ç”¨äºé•¿åŸºå› ï¼Œé¿å…ä¿¡æ¯ä¸¢å¤±|

<br>

---
## Profile of peaks binding to TSS

### TagMatrix

- `TagMatrix`: peaksåœ¨åŸºå› ç»„åŒºåŸŸå†…çš„ä½ç½®åˆ†å¸ƒï¼Œå¹¶ä¸ºæ¯ä¸ªåŒºåŸŸ(bins)ç”Ÿæˆä¸€ä¸ªçŸ©é˜µ
  
    * `TagMatrix` çš„**è¡Œ**ä»£è¡¨ **ä¸åŒçš„peaks**ã€‚
    
    * `TagMatrix` çš„**åˆ—**ä»£è¡¨**åŒºåŸŸè¢«åˆ’åˆ†çš„ bins**
    <br>
- å‡è®¾ä½ å¯¹åŸºå› çš„ **è½¬å½•èµ·å§‹ä½ç‚¹ï¼ˆTSSï¼‰** è¿›è¡Œåˆ†æï¼Œå¹¶è®¾å®š `upstream = 3000`ã€`downstream = 3000`ï¼Œä¸” `nbin = 800`ï¼š
  
    - è¿™ **6000 bp** è¢«åˆ’åˆ†ä¸º **800 ä¸ª bin**ï¼Œå³ **æ¯ä¸ª bin ä»£è¡¨ 6000/800 = 7.5 bp** çš„åŒºé—´ã€‚
      
    - `TagMatrix[i, j]` ä»£è¡¨ ç¬¬ i ä¸ª peak åœ¨ç¬¬ j ä¸ª bin é‡Œçš„ä¿¡å·å€¼ï¼ˆå¦‚è¯»å–æ•°æˆ–åŠ æƒå€¼)
    <br>   
- `getTagMatrix()` can get `TagMatrix`from `GRanges` object
```r
tagMatrix <- getTagMatrix(
  peak_list[[1]],       # peak file or GRanges object
  weightCol = "score",  # column name of peak as weight
  TxDb = TxDb,          # TxDb object
  upstream = 500,
  downstream = 500,
  verbose = TRUE,       # print message or not
  ignore_strand = FALSE,# æ˜¯å¦å¿½ç•¥é“¾ä¿¡æ¯(é»˜è®¤åŒºåˆ†æ­£è´Ÿé“¾)
  
  nbin = 500,          # window region è¢«åˆ’åˆ†æˆå¤šå°‘ä¸ª binsï¼ˆå°åŒºé—´ï¼‰ï¼ˆç”¨äºè®¡ç®—ä¿¡å·å¯†åº¦ï¼‰
                       # å€¼è¶Šå¤§ï¼Œåˆ†è¾¨ç‡è¶Šé«˜ï¼Œä½†å¯èƒ½ä¼šå¯¼è‡´è®¡ç®—æ—¶é—´å¢åŠ 
  
  
  #windows,            # æ¥æä¾›ä¸€ä¸ªåŸºå› ç»„åŒºåŸŸ(window region)çš„ GRanges å¯¹è±¡
                       # é€‚ç”¨äºåˆ†ææŸä¸ªåŸºå› çš„å¤–æ˜¾å­ã€å†…å«å­ç­‰ç‰¹å®šwindow regionçš„å³°å€¼åˆ†å¸ƒ
                       # å¦‚æœæä¾›äº† windows,é‚£ä¹ˆ upstream å’Œ downstream å‚æ•°ä¼šè¢«å¿½ç•¥ï¼‰
  
  type = "start_site", # é€‰æ‹©ç»˜åˆ¶åŒºåŸŸ("start_site"ï¼šTSS, Transcription Start Siteï¼Œ
                       #               "end_site"ï¼šTES, Transcription End Site
                       #               "body" : Gene Body (æ•´ä¸ªåŸºå› ä½“))
  
  by = "gene"         # æŒ‰'gene', 'transcript', 'exon', 'intron' , '3UTR' , '5UTR', 'UTR'è¿›è¡Œåˆ†æ
)
```

<br>

* `TagMatrix` structure
```r
tagMatrix[c(1:10),c(1:10)]
```

```text
       [,1]    [,2]    [,3]    [,4]    [,5]    [,6]    [,7]    [,8]    [,9]   [,10]
 [1,] 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000
 [2,] 5.18626 5.18626 5.18626 5.18626 5.18626 5.18626 5.18626 5.18626 5.18626 5.18626
 [3,] 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000
 [4,] 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000
 [5,] 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000
 [6,] 5.17740 5.17740 5.17740 5.17740 5.17740 5.17740 5.17740 5.17740 5.17740 5.17740
 [7,] 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000
 [8,] 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000
 [9,] 6.33876 6.33876 6.33876 6.33876 6.33876 6.33876 6.33876 6.33876 6.33876 6.33876
[10,] 0.00000 0.00000 3.35933 6.71866 6.71866 6.71866 6.71866 6.71866 6.71866 6.71866
```

<br>

### Heatmap and peak profiling
* you can use `tagHeatmap()` to create heatmap with `TagMatrix`

	```r
	tagHeatmap(tagMatrix)
	```
	<img src="./picture/019.jpg" width="600"/>

<br>

* you can use `peakHeatmap()` to create heatmap with  peak file or `GRanges` object

	```r
	peakHeatmap(
	  peak_list[[1]],    # peak file or GRanges object
	  weightCol = "score",  # column name of peak as weight
	  TxDb = TxDb,        # TxDb object
	  upstream = 500,     # upstream position  ,rel(0.2)
	  downstream = 500,   # downstream position
	  xlab = "",         # xlab
	  ylab = "",         # ylab
	  title = NULL,      
	  palette = NULL,    # é¢œè‰²æ–¹æ¡ˆ
	  verbose = FALSE,   # print message or not
	  by = "gene",       # æŒ‰'gene', 'transcript', 'exon', 'intron' , '3UTR' , '5UTR', 'UTR'è¿›è¡Œåˆ†æ
	  
	  type = "start_site", # é€‰æ‹©ç»˜åˆ¶åŒºåŸŸ("start_site"ï¼šTSS, Transcription Start Siteï¼Œ
	                       #        "end_site"ï¼šTES, Transcription End Site
	                       #        "body" : Gene Body (æ•´ä¸ªåŸºå› ä½“))
	  
	  nbin = 500,          # window region è¢«åˆ’åˆ†æˆå¤šå°‘ä¸ª binsï¼ˆå°åŒºé—´ï¼‰ï¼ˆç”¨äºè®¡ç®—ä¿¡å·å¯†åº¦ï¼‰
	                       # å€¼è¶Šå¤§ï¼Œåˆ†è¾¨ç‡è¶Šé«˜ï¼Œä½†å¯èƒ½ä¼šå¯¼è‡´è®¡ç®—æ—¶é—´å¢åŠ 
	  
	  ignore_strand = FALSE, # æ˜¯å¦å¿½ç•¥é“¾ä¿¡æ¯ï¼ˆé»˜è®¤åŒºåˆ†æ­£è´Ÿé“¾ï¼‰
	  
	  #windows,           # æ¥æä¾›ä¸€ä¸ªåŸºå› ç»„åŒºåŸŸ(window region)çš„ GRanges å¯¹è±¡
	                      # é€‚ç”¨äºåˆ†ææŸä¸ªåŸºå› çš„å¤–æ˜¾å­ã€å†…å«å­ç­‰ç‰¹å®šwindow regionçš„å³°å€¼åˆ†å¸ƒ
	                      # å¦‚æœæä¾›äº† windows,é‚£ä¹ˆ upstream å’Œ downstream å‚æ•°ä¼šè¢«å¿½ç•¥ï¼‰


â€‹	  
	  ncol = NULL,       # æŒ‡å®šè¾“å‡ºçš„å›¾åƒåˆ—æ•°
	  nrow = NULL        # æŒ‡å®šè¾“å‡ºçš„å›¾åƒè¡Œæ•°
	)
	```
	<img src="./picture/020.jpg" width="600"/>

<br>

* `peakHeatmap_multiple_Sets()` can compare of the peak heatmap over **two regions**

	```r
	#æå–æ‰€æœ‰è½¬å½•æœ¬çš„åŸºå› ä¿¡æ¯
	txdb1 <- transcripts(TxDb)
	
	#ä»…é€‰æ‹©å‰ 10,000 ä¸ªåŸºå› çš„ 5' UTRï¼Œé¿å…è®¡ç®—é‡è¿‡å¤§
	txdb2 <- unlist(fiveUTRsByTranscript(TxDb))[1:10000,]
	
	region_list <- list(geneX = txdb1, geneY = txdb2)
	
	peakHeatmap_multiple_Sets(peak_list[[1]],
	                          upstream = 500,
	                          downstream = 500,
	                          by = c("geneX","geneY"),
	                          type = "start_site",
	                          TxDb = region_list,
	                          nbin = 500,
	                          verbose = FALSE)
	```
	<img src="./picture/021.jpg" width="600"/>

<br>

- `peak_Profile_Heatmap()` can plot heatmap and peak profiling together

	```r
	peak_Profile_Heatmap(peak = peak_list[[1]],
	                     upstream = 500,
	                     downstream = 500,
	                     by = "gene",
	                     type = "start_site",
	                     TxDb = TxDb,
	                     nbin = 500,
	                     verbose = FALSE)
	```
	<img src="./picture/022.jpg" width="600"/>

<br>

- `peak_Profile_Heatmap()` is also support exploring several regions with heatmap and peak profiling

	```r
	#æå–æ‰€æœ‰è½¬å½•æœ¬çš„åŸºå› ä¿¡æ¯
	txdb1 <- transcripts(TxDb)
	
	#ä»…é€‰æ‹©å‰ 10,000 ä¸ªåŸºå› çš„ 5' UTRï¼Œé¿å…è®¡ç®—é‡è¿‡å¤§
	txdb2 <- unlist(fiveUTRsByTranscript(TxDb))[1:10000,]
	
	region_list <- list(geneX = txdb1, geneY = txdb2)


â€‹	
	peak_Profile_Heatmap(peak = peak_list[[1]],
	                     upstream = 500,
	                     downstream = 500,
	                     by = c("geneX","geneY"),
	                     type = "start_site",
	                     TxDb = region_list,
	                     nbin = 500,
	                     verbose = FALSE)
	```
	<img src="./picture/023.jpg" width="600"/>

<br>

### Average Profile of peaks binding to TSS region

- `plotAvgProf()` can draw Average Profile of peaks with `TagMatrix`
  
	- **parametets:`conf`(confidence interval): å›¾ä¸­ç°è‰²çš„åœ°æ–¹ï¼Œ`conf = 0.95` è¡¨ç¤º æœŸæœ›95%çš„æ ·æœ¬ç»“æœè½å…¥è¿™ä¸ªåŒºé—´

	```r
	plotAvgProf(tagMatrix, 
	            xlim=c(1, 500),   # Need to be smaller than and match your nbin 
	            xlab="Genomic Region (5'->3')", 
	            ylab = "Read Count Frequency" ,
	            conf = 0.95)
	```
	<img src="./picture/024.jpg" width="600"/>

<br>

- `plotAvgProf2()` can draw Average Profile of peaks with **peak file or `GRanges` object**

	```r
	plotAvgProf2(
	  peak_list[[1]],    # peak file or GRanges object
	  weightCol = NULL,  # column name of peak as weight
	  TxDb = TxDb,
	  upstream = 500,
	  downstream = 500,
	  xlab = "Genomic Region (5'->3')",
	  ylab = "Peak Count Frequency",
	  conf = 0.95,        # confidence interval
	  facet = "none",     # åˆ†ç»„ï¼š"none"ï¼ˆé»˜è®¤ï¼‰ï¼šä¸åˆ†ç»„,"row"ï¼šæŒ‰è¡Œåˆ†ç»„,"column"ï¼šæŒ‰åˆ—åˆ†ç»„
	  free_y = TRUE,      # æ˜¯å¦å…è®¸ Y è½´è‡ªç”±ç¼©æ”¾  
	  verbose = FALSE,    #  print message or not
	  ignore_strand = FALSE,
	)
	```
	<img src="./picture/025.jpg" width="600"/>

<br>

---
## peak annotation

- `annotatePeak()` is use to annotate the peaks

	```r
	{r}
	peakAnno <-annotatePeak(
	  peak_list[[1]],              # Input peak data (GRanges object)
	  tssRegion = c(-3000, 3000),  # Region around TSS to annotate 
	  TxDb = TxDb,                 # Annotation database (TxDb object)
	  level = "transcript",        # Annotation level ("gene", "transcript")
	  
	  assignGenomicAnnotation = TRUE,  # annotationåˆ—æ˜¯å¦ä¼šæŠ¥å‘Šè¯¦ç»†ä¿¡æ¯
	  
	  # Order of annotation features
	  # å®šä¹‰æ³¨é‡Šçš„ä¼˜å…ˆçº§ï¼Œå³å¦‚æœä¸€ä¸ª peak å½’å±äºå¤šä¸ªåŒºåŸŸï¼Œä¼˜å…ˆæŒ‰è¿™ä¸ªé¡ºåºåˆ†é…
	  genomicAnnotationPriority = c("Promoter", "5UTR", "3UTR", "Exon", "Intron", "Downstream", "Intergenic"),
	  
	  # External annotation database (optional)
	  annoDb = "org.Rn.eg.db", 
	  
	  # Whether to include flanking gene regions
	  # if addFlankGeneInfo is TRUEï¼ŒThe program will find the nearest gene within a certain distance (flankDistance) to the peak.
	  addFlankGeneInfo = FALSE,  
	  flankDistance = 5000,  


â€‹	  
	  sameStrand = FALSE,  # Consider only the same strand as the gene
	  ignoreOverlap = FALSE,  #  whether ignore overlap of TSS with peak
	  ignoreUpstream = FALSE,  # Ignore upstream regions
	  ignoreDownstream = FALSE,  # Ignore downstream regions
	  
	  # Type of overlap ("TSS" or "all")
	  # overlap = "TSS":(deafult),When the peak covers TSS (transcription start site), the gene is considered to be the nearest gene.
	  # overlap = "all":If the peak falls in any gene region, the gene will be directly considered as the nearest gene without additionally searching for the nearest TSS
	  overlap = "TSS",  
	  verbose = TRUE,  # Whether to print progress messages
	  
	  # Specify the gene information to be output 
	  # (e.g: "ENTREZID", "ENSEMBL", "SYMBOL", "GENENAME")
	  columns = c("ENTREZID", "SYMBOL", "GENENAME")
	)
	```

<br>

- `annotatePeak()` results in `csAnno`ï¼ˆChIPseeker Annotation)
  
    - `as.GRanges()`: convert to `GRanges` objectsã€‚
      
    - `as.data.frame()`: convert to `data.frame`

	```r
	peakAnno_dataframe <- as.data.frame(peakAnno)
	
	peakAnno_dataframe[c(1:6),]
	```

	```text
	seqnames start    end      width strand score   annotation                                     geneChr geneStart geneEnd  geneLength geneStrand geneId transcriptId distanceToTSS SYMBOL   GENENAME
	chr1     1193628  1193990  363   *      3.49474 Intron (NM_001013063/292461, intron 5 of 6)    1       1101965   1116926  14962      2          679825 NM_001161691 -76702        Raet1ll1 retinoic acid early transcript 1L like 1
	chr1     1670780  1671341  562   *      6.46071 Distal Intergenic                              1       1702696   1731210  28515      1          292462 NM_001106217 -31355        Lrp11    LDL receptor related protein 11
	chr1     1784527  1784797  271   *      4.72403 Promoter (<=1kb)                               1       1784078   1817310  33233      1          308265 NM_001134543 449           Lats1    large tumor suppressor kinase 1
	chr1     1803427  1803932  506   *      3.62225 Exon (NM_001134543/308265, exon 4 of 8)        1       1784078   1817310  33233      1          308265 NM_001134543 19349         Lats1    large tumor suppressor kinase 1
	chr1     1925707  1926101  395   *      2.62105 Intron (NM_001108457/361449, intron 12 of 12)  1       1938419   1972918  34500      1          308266 NM_001107469 -12318        Zc3h12d  zinc finger CCCH type containing 12D
	chr1     2352454  2352970  517   *      2.62105 Intron (NM_001108458/361450, intron 7 of 7)    1       2333823   2627475  293653     2          361450 NM_001108458 274505        Ust      uronyl-2-sulfotransferase
	```

<br>

å½“æ³¨é‡Šä¸º**Exon**æˆ–**Intron**æ—¶ï¼Œä¸”`assignGenomicAnnotation = TRUE`æ—¶ï¼Œ`annotatePeak()` è¿”å›çš„`annotation`åˆ—ä¼šæŠ¥å‘Šè¯¦ç»†ä¿¡æ¯ã€‚

- ä¾‹å¦‚ï¼ŒExon (NM_001134543/308265, exon 4 of 8)
  
    * è¡¨ç¤ºè¯¥peakä¸è½¬å½•æœ¬ **NM_001134543** çš„å¤–æ˜¾å­é‡å 
    
    * å¯¹åº”çš„**Entrez ID**ä¸º**308265**
    
    * è¿™ä¸ªé‡å çš„å¤–æ˜¾å­æ˜¯è¯¥è½¬å½•æœ¬**NM_001134543** çš„ç¬¬4ä¸ªå¤–æ˜¾å­

<br>

---
## Visualize Genomic Annotation

### Pie plot
```r
plotAnnoPie(peakAnno)
```
<img src="./picture/026.jpg" width="600"/>

<br>

### Bar plot
```r
plotAnnoBar(peakAnno)
```
<img src="./picture/027.jpg" width="600"/>

<br>

### Ven pie plot
User may interested to view the full annotation with their overlap
```r
vennpie(peakAnno)
```
<img src="./picture/028.jpg" width="600"/>

<br>

### Upsetplot()

We extend `UpSetR` to view full annotation overlap. User can user `upsetplot()` function.
```r
upsetplot(peakAnno)
```
<img src="./picture/029.jpg" width="600"/>

<br>

We can combine `vennpie` with `upsetplot()` by setting `vennpie = TRUE`.
```r
upsetplot(peakAnno, vennpie=TRUE)
```
<img src="./picture/030.jpg" width="600"/>

<br>

### Visualize distribution of TF-binding loci relative to TSS
```r
plotDistToTSS(peakAnno,
              title="Distribution of transcription factor-binding loci\nrelative to TSS")
```
<img src="./picture/031.jpg" width="600"/>

**X è½´**ï¼šä»£è¡¨ç»“åˆä½ç‚¹çš„ç™¾åˆ†æ¯”ï¼ˆBinding sites (%)ï¼‰ï¼Œå³ä¸åŒåŒºé—´å†…çš„ TF ç»“åˆä½ç‚¹å æ¯”

<br>

---

## Functional enrichment analysis

- Detailed enrichment analysis process see **[clusterProfiler](http://bioconductor.org/packages/clusterProfiler)**
    <br>
- you can use gene set of `csAnno` object as input in enrichment analysis

	```r
	library(ReactomePA)
	pathway1 <- enrichPathway(as.data.frame(peakAnno)$geneId)
	```

<br>

- [ChIPseeker](http://bioconductor.org/packages/ChIPseeker) also provides `seq2gene` function, user can use `GRanges` objects as input to get gene set (**Entrez gene ID**)
  
    You can obtain **different genes** by mapping **different associations between gene regions**

	```r
	gene <- seq2gene(peak_list[[1]],                # GRanges object
	                 tssRegion = c(-1000, 1000),    # TSS region
	                 flankDistance = 3000,          # flanking search radius  
	                 TxDb=TxDb,                     # TxDb object
	                 sameStrand = FALSE)            # logical whether find  gene in the same strand
	
	gene[1:20]
	```

```text
 [1] "308265" "361450" "292640" "114490" "308632" "116465" "308923" "293947" "685940" "309565"
[11] "64032"  "65139"  "29517"  "361462" "246769" "498999" "292206" "361191" "290946" "361472"
```

<br>

---
## peak data set comparison

### Profile of several peak data binding to TSS region

- `plotAvgProf()`, `tagHeatmap()` and `plotPeakProf()` can accept a list of `tagMatrix` and visualize profile or heatmap among several Peaks
    <br>
- `plotAvgProf2()`, `peakHeatmap()` and `plotPeakProf2()` can accept a list of **bed files or `GRange` list** and perform the same task
    <br>
- the above function usage see
	- ğŸª¨[[#Heatmap and peak profiling]]   ğŸ”—[Heatmap and peak profiling](#Heatmap-and-peak-profiling)
	- ğŸª¨[[#Average Profile of peaks binding to TSS region]]  ğŸ”—[Average Profile of peaks binding to TSS region](#Average-Profile-of-peaks-binding-to-TSS-region)

<br>

```r
plotPeakProf2(peak_list, upstream = 3000, downstream = 3000, conf = 0.95,
              by = "gene", type = "start_site", TxDb = TxDb,
              facet = "row", nbin = 800)
```
<img src="./picture/032.jpg" width="600"/>

<br>

### peak annotation comparision

 The `plotAnnoBar()` and `plotDistToTSS()` can also accept input of a **named list of annotated peaks**

```r
# Use lapply to process each element
peakAnnoList <- lapply(peak_list, annotatePeak, TxDb=TxDb,
                       tssRegion=c(-3000, 3000), verbose=FALSE)
```

```r
plotAnnoBar(peakAnnoList)
```
<img src="./picture/033.jpg" width="600"/>

<br>

```r
plotDistToTSS(peakAnnoList)
```
<img src="./picture/034.jpg" width="600"/>

<br>

### Functional profiles comparison

The **[clusterProfiler](http://bioconductor.org/packages/clusterProfiler)** package provides `compareCluster()` function for comparing biological themes among gene clusters, and can be easily adopted to compare different peak experiments.
```r
# Use lapply to process each element to get gene set
genes = lapply(peakAnnoList, function(i) as.data.frame(i)$geneId)

# enrichment analysis
compKEGG <- compareCluster(geneCluster   = genes,
                           fun           = "enrichKEGG",
                           pvalueCutoff  = 0.05,
                           pAdjustMethod = "BH",
                           organism = "rno",)

# dotplot
dotplot(compKEGG, showCategory = 15, title = "KEGG Pathway Enrichment Analysis")
```


<br>

### Overlap of peaks and annotated genes

- `vennplot` accept a list of object, can be a **list of GRanges** or a **list of vector**.
  
- The detailed **Overlap of peaks** you should see ğŸª¨ [[#Peak overlap]] ğŸ”— [Peak overlap](#Peak-overlap)

```r
# a list of GRanges
vennplot(peak_list)
```
<img src="./picture/035.jpg" width="600"/>

```r
# a list of vector
genes= lapply(peakAnnoList, function(i) as.data.frame(i)$geneId)
vennplot(genes)
```
<img src="./picture/036.jpg" width="600"/>

<br>

---
## ChIP-seq Overlap Statistical testing

Overlap is very important, if two ChIP experiment by **two different proteins** **overlap in a large fraction** of their peaks, **they may cooperative in regulation.**


### Shuffle genome coordination

- The `shuffle()` function to randomly permute the genomic locations of ChIP peaks defined in a genome which stored in `TxDb` object
  
- `shuffle()` **éšæœºåŒ– ChIP-seq å³°çš„åŸºå› ç»„åæ ‡**ï¼Œä½†ç¡®ä¿å®ƒä»¬ä»ç„¶ä½äº **åŸºå› ç»„çš„æœ‰æ•ˆåŒºåŸŸ**ã€‚è¿™ä¸ªæ–¹æ³•å¸¸ç”¨äºè®¡ç®— **èƒŒæ™¯åˆ†å¸ƒ**ï¼Œè¿›è€Œè¯„ä¼° ChIP-seq å³°çš„ **å¯Œé›†æ˜¾è‘—æ€§**

```r
p <- GRanges(seqnames=c("chr1", "chr3"),
             ranges=IRanges(start=c(1, 100), end=c(50, 130)))

shuffle(p, TxDb=TxDb)
```

```text
GRanges object with 2 ranges and 0 metadata columns:
      seqnames              ranges strand
         <Rle>           <IRanges>  <Rle>
  [1]     chr1 241954092-241954141      *
  [2]     chr3 161021597-161021627      *
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths
```

<br>

### Peak overlap enrichment analysis

`enrichPeakOverlap()` å·¥ä½œåŸç†**

1. **è®¡ç®—çœŸå®é‡å å³°å€¼æ•°**
   
    - è®¡ç®— `queryPeak`å’Œ `targetPeak`ä¹‹é—´çš„å®é™…é‡å å³°æ•°é‡ã€‚
    <br>
2. **ç”ŸæˆéšæœºèƒŒæ™¯åˆ†å¸ƒ**
   
    - ä½¿ç”¨ `shuffle()` å‡½æ•° **éšæœºåŒ–** `queryPeak` çš„åŸºå› ç»„åæ ‡ï¼Œç”Ÿæˆ **nShuffle ç»„** éšæœºå³°æ•°æ®ã€‚
    
    - å¯¹æ¯ç»„éšæœºå³°é›†ï¼Œä¸ `targetPeak` è®¡ç®—é‡å å³°å€¼æ•°ï¼Œä»è€Œæ„å»º **èƒŒæ™¯åˆ†å¸ƒï¼ˆnull distributionï¼‰**ã€‚
    <br>
3. **è®¡ç®—å¯Œé›†æ˜¾è‘—æ€§**
   
    - ç»Ÿè®¡å®é™…å³°å€¼é‡å æ•° **æ˜¯å¦è¿œé«˜äºéšæœºèƒŒæ™¯**ï¼Œç”¨ **p å€¼** è¯„ä¼°å…¶æ˜¾è‘—æ€§ã€‚
    
    - é‡‡ç”¨ Benjamini-Hochbergï¼ˆ`BH`ï¼‰æ–¹æ³•è¿›è¡Œ **å¤šé‡å‡è®¾æ£€éªŒæ ¡æ­£**æ¥æ ¡æ­£ p å€¼ã€‚

<br>

```r
Overlap_analysis=enrichPeakOverlap(
  queryPeak=peak_list[[1]],       # bed file or GRanges object
  targetPeak=peak_list,  # target bed file(s) or folder or a list of GRanges objects
  TxDb= TxDb,           # TxDb objects
  pAdjustMethod="BH",   # pvalue adjustment method
  nShuffle=50,          # shuffle numbers
  chainFile=NULL,       # chain file for liftOver(åŸºå› è½´åæ ‡è½¬æ¢)
  pool=FALSE,           # whether Merge targetPeak (multiple ChIP seq peaks)
  mc.cores = detectCores()-1,  # number of cores
  verbose= FALSE)  # whether Print detailed information  
```

- **Error**: The input of `list of GRanges` for `targetPeak` here will result in **different target peak p-values being the same**. If we input `bed file (s) or folder` , this problem will not occur.
    <br>
- To speed up the compilation of this vignettes, we only set `nShuffle` to **50** as an example for only demonstration. User should set the number to **1000 or above** for more robust result
    <br>
- Parameter `chainFile` are chain file name for mapping the `targetPeak` to the genome version consistent with `queryPeak` when their genome version are different.
  
    - This creat the possibility of comparison among different genome version and cross species
      
    - The detailed `chainFile` you should see ğŸª¨[[#Chain file]]  ğŸ”—[Chain file](#Chain-file)

<br>

```r
Overlap_analysis
```

```text
           qSample  tSample  qLen  tLen  N_OL     pvalue   p.adjust
peak0625 queryPeak peak0625 10714 10714 10714 0.01960784 0.01960784
peak1011 queryPeak peak1011 10714 10019  1833 0.01960784 0.01960784
```

- `qSample` is the name of `queryPeak`
  
- `qLen` is the the number of peaks in `queryPeak`
  
- `N_OL` is the number of overlap between `queryPeak` and `targetPeak`

<br>

---
## ChIP-seq Data Mining from GEO

### GEO data collection

There are many ChIP seq data sets that have been published and deposited in GEO database

* [ChIPseeker](http://bioconductor.org/packages/ChIPseeker)collect about **17,000** **bed files** deposited in GEO

- User can use `getGEOspecies()` to get a summary based on speices

	```r
	GEOspecies <-getGEOspecies()
	GEOspecies[c(1:10),]
	```
	
	```text
	                                               species Freq
	1                      Actinobacillus pleuropneumoniae    1
	2  Actinobacillus pleuropneumoniae serovar 3 str. JL03    1
	3                                        Aedes aegypti   11
	4                                             Anabaena    6
	5                                  Anolis carolinensis    7
	6                                    Anopheles gambiae    2
	7                                       Apis mellifera   17
	8                            Apis mellifera scutellata    1
	9                                          Arabidopsis   17
	10                                  Arabidopsis lyrata    4
	```


<br>

* User can use `getGEOgenomeVersion()` to get a summary based on genome version

	```r
	GEOgenomeVersion <- getGEOgenomeVersion()
	GEOgenomeVersion
	```
	
	```text
	                        organism genomeVersion  Freq
	1            Anolis carolinensis       anoCar2     7
	2                     Bos taurus       bosTau4     2
	3                     Bos taurus       bosTau6    33
	4                     Bos taurus       bosTau7     2
	5         Canis lupus familiaris       canFam3    10
	6         Caenorhabditis elegans          ce10   328
	7         Caenorhabditis elegans           ce6    64
	8                    Danio rerio       danRer6    89
	9                    Danio rerio       danRer7    89
	10       Drosophila melanogaster           dm3   767
	11                 Gallus gallus       galGal3    33
	12                 Gallus gallus       galGal4    73
	13        Gasterosteus aculeatus       gasAcu1     1
	14         Heterocephalus glaber       hetGla2     4
	15                  Homo sapiens          hg18  3368
	16                  Homo sapiens          hg19 30326
	17                  Homo sapiens          hg38  4091
	18                  Mus musculus          mm10 19089
	19                  Mus musculus           mm8   556
	20                  Mus musculus           mm9 17247
	21         Monodelphis domestica       monDom5    31
	22                    Ovis aries       oviAri3     2
	23               Pan troglodytes       panTro3    48
	24               Pan troglodytes       panTro4    48
	25                Macaca mulatta       rheMac2    84
	26                Macaca mulatta       rheMac3    40
	27             Rattus norvegicus           rn5    66
	28      Saccharomyces cerevisiae       sacCer2   145
	29      Saccharomyces cerevisiae       sacCer3  2646
	30                    Sus scrofa       susScr2    17
	31                    Sus scrofa       susScr3    18
	32           Taeniopygia guttata       taeGut2    20
	33 Xenopus (Silurana) tropicalis       xenTro3     3
	```

<br>

* User can access the detail information by `getGEOInfo()`, for each genome version.

	```r
	rn5 <- getGEOInfo(genome="rn5", simplify=TRUE)
	rn5[c(1:6),]
	```
	
	```text
	series_id          gsm         organism           title               supplementary_file                                                                                           genomeVersion pubmed_id
	GSE61204,GSE61207  GSM1499535  Rattus norvegicus  MEF2A_ChIP-exo 1   ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1499nnn/GSM1499535/suppl/GSM1499535_mef2_rn5_q37_nodups.nomodel.tolarge_peaks.bed.gz  rn5            <NA>
	GSE72725,GSE72727  GSM1869162  Rattus norvegicus  Chd7_d3            ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1869nnn/GSM1869162/suppl/GSM1869162_Chd7_d3_peaks.bed.gz                             rn5            <NA>
	GSE72725,GSE72727  GSM1869163  Rattus norvegicus  Sox10_d3           ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1869nnn/GSM1869163/suppl/GSM1869163_Sox10_d3_peaks.bed.gz                            rn5            <NA>
	GSE100505          GSM2685495  Rattus norvegicus  RAT_EDL_M3         ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM2685nnn/GSM2685495/suppl/GSM2685495_junctions_all_RAT_EDL_M3.bed.gz                  rn5            <NA>
	GSE100505          GSM2685514  Rattus norvegicus  RAT_SOL_F4         ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM2685nnn/GSM2685514/suppl/GSM2685514_junctions_all_RAT_SOL_F4.bed.gz                  rn5            <NA>
	GSE100505          GSM2685493  Rattus norvegicus  RAT_EDL_M1         ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM2685nnn/GSM2685493/suppl/GSM2685493_junctions_all_RAT_EDL_M1.bed.gz                  rn5            <NA>
	```

<br>

### Download GEO ChIP data sets

[ChIPseeker](http://bioconductor.org/packages/ChIPseeker) provide function `downloadGEObedFiles()` to download **all the bed files of a particular genome**

```r
# creat rn5 folder
dir.create("rn5", showWarnings = FALSE, recursive = TRUE)  
```

```r
# downland geo bed file
downloadGEObedFiles(genome="rn5", # genome version
                    destDir="rn5"  # #downland location
                    )
```

<br>

è¿™é‡Œä½œè€…æä¾›çš„`downloadGEObedFiles()` ä»£ç **æœ‰é—®é¢˜**ï¼Œå› ä¸ºGEO æœåŠ¡å™¨å¯èƒ½å·²ç»ä¸æ”¯æŒ FTP è®¿é—®ï¼Œå»ºè®®æ”¹ç”¨ HTTPSï¼Œæ‰€ä»¥æˆ‘å¯¹ä»£ç è¿›è¡Œäº†ä¿®æ”¹ï¼Œåˆ›å»ºäº†ä¸ªæ–°å‡½æ•°`downloadGEObedFilesHTTPS()`
```r
downloadGEObedFilesHTTPS <- function(genome,destDir) {
    # Obtaining genomic information
    info <- getGEOInfo(genome)
    
    # Get supplementary_file column
    fnames <- as.character(info$supplementary_file)
    
    # Replace FTP with HTTPS
    fnames_https <- gsub("^ftp://", "https://", fnames)
    
    # Generate target file path
    destfiles <- file.path(destDir, basename(fnames_https))
    names(destfiles) <- NULL
    
    # Traverse and download files
    for (i in seq_along(fnames_https)) {
        
      # æ£€æŸ¥ destfiles[i]æ˜¯å¦å·²å­˜åœ¨ï¼Œæ–‡ä»¶ä¸å­˜åœ¨ï¼Œåˆ™æ‰§è¡Œä¸‹è½½
      if (!file.exists(destfiles[i])) {
            tryCatch(
                download.file(fnames_https[i], 
                              destfile = destfiles[i], 
                              mode = "wb"),#ä»¥äºŒè¿›åˆ¶æ¨¡å¼ä¸‹è½½ï¼Œç¡®ä¿æ–‡ä»¶ä¸ä¼šå› æ¢è¡Œç¬¦é—®é¢˜æŸåã€‚
                error = function(e) message(fnames_https[i], ": file not found and skip")
            )
        }
    }
}
```

```r
downloadGEObedFilesHTTPS(genome="rn5", destDir="rn5")
```

<br>

### Overlap significant testing

- After download the bed files from GEO, we can pass them to for testing the significant of overlap
    <br>
- **Signifcant overlap** can be use to generate hypothesis of cooperative regulation
    <br>
- We can use `enrichPeakOverlap()` function to assess **Statistical testing** . detail see ğŸª¨[[#ChIP-seq Overlap Statistical testing]]  ğŸ”—[ChIP-seq Overlap Statistical testing](#ChIP-seq-Overlap-Statistical-testing)
    <br>
- In different genome versions, parameter `chainFile` that can pass a `chain file` and `liftOver` the `targetPeak` to the genome version consistent with `queryPeak`